core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -1281,13 +1349,314 @@ public String toStringResponseOnly(Function<? super ResponseHeaders, ?> headersS
         return sanitized != null ? sanitized.toString() : "<sanitized>";
     }
 
-    private static final class ListenerEntry {
-        final RequestLogListener listener;
+    private static final class RequestLogFuture extends CompletableFuture<RequestLog> {
         final int interestedFlags;
 
-        ListenerEntry(RequestLogListener listener, int interestedFlags) {
-            this.listener = listener;
+        RequestLogFuture(int interestedFlags) {
             this.interestedFlags = interestedFlags;
         }
+
+        void completeLog(RequestLog log) {
+            super.complete(log);
+        }
+
+        @Override
+        public boolean complete(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public void obtrudeValue(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void obtrudeException(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    /**
+     * A {@link RequestLog} that delegates to {@link DefaultRequestLog} to read properties without
+     * checking availability.
+     */
+    private final class CompleteRequestLog implements RequestLog {
+        @Override
+        public boolean isComplete() {
+            return DefaultRequestLog.this.isComplete();
+        }
+
+        @Override
+        public boolean isRequestComplete() {
+            return DefaultRequestLog.this.isRequestComplete();
+        }
+
+        @Override
+        public boolean isAvailable(RequestLogProperty property) {
+            return DefaultRequestLog.this.isAvailable(property);
+        }
+
+        @Override
+        public RequestLog partial() {
+            return DefaultRequestLog.this.partial();
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> completeFuture() {
+            return DefaultRequestLog.this.completeFuture();
```

Comment:
I wonder if there's any advantage to returning a new completed future GC-wise / escape-analysis-wise for all of these. Though I guess they're pretty rarely used anyways

Reply:
Will use `UnmodifiableFuture.completedFuture(null)` which returns a singleton once #2306 is merged.